### ISSUE TITLE
Wrong type recognised

### ISSUE THREAD

**Issue by ekulabuhov** [microsoft/TypeScript#29773]
**TypeScript Version:**  3.4.0-dev.20190206

**Search Terms:**
mapbox, type resolution

**Code**

```ts
const usa_bbox = [
  [-126.12982731363905, 23.011472701841583],
  [-67.14927405043929, 50.54700547505038]
];


type LngLatLike = { lng: number; lat: number; } | { lon: number; lat: number; } | [number, number];
type LngLatBoundsLike = [LngLatLike, LngLatLike] | [number, number, number, number];   

function cameraForBounds(bounds: LngLatBoundsLike) {}

cameraForBounds(usa_bbox); // <-- Argument of type 'number[][]' is not assignable to parameter of type 'LngLatBoundsLike'. Type 'number[][]' is missing the following properties from type '[number, number, number, number]': 0, 1, 2, 3
```

**Expected behavior:**
It shouldn't show an error. Should detect type: [LngLatLike, LngLatLike]

**Actual behavior:**
Throws an error for the incorrect type.

**Playground Link:** 
https://www.typescriptlang.org/play/index.html#src=const%20usa_bbox%20%3D%20%5B%0D%0A%20%20%5B-126.12982731363905%2C%2023.011472701841583%5D%2C%0D%0A%20%20%5B-67.14927405043929%2C%2050.54700547505038%5D%0D%0A%5D%3B%0D%0A%0D%0A%0D%0Atype%20LngLatLike%20%3D%20%7B%20lng%3A%20number%3B%20lat%3A%20number%3B%20%7D%20%7C%20%7B%20lon%3A%20number%3B%20lat%3A%20number%3B%20%7D%20%7C%20%5Bnumber%2C%20number%5D%3B%0D%0Atype%20LngLatBoundsLike%20%3D%20%5BLngLatLike%2C%20LngLatLike%5D%20%7C%20%5Bnumber%2C%20number%2C%20number%2C%20number%5D%3B%20%20%20%0D%0A%0D%0Afunction%20cameraForBounds(bounds%3A%20LngLatBoundsLike)%20%7B%7D%0D%0A%0D%0AcameraForBounds(usa_bbox)%3B

**Related Issues:**

**Comment by dragomirtitian** [comment]
I literally answered this exact question on SO today [link](https://stackoverflow.com/questions/54557317/write-an-object-with-keys-and-array-typescript/54557421#54557421)

Long story short, typescript will not infer tuples for array literals unless there is a reason to. For  `usa_bbox`  there is no reason to, so it will be `number[][]` which will not be assignable to the tuple type.

In 3.4 you will be able to use `as const` to make he compiler infer a readonly tuple:

```ts
const usa_bbox = [
  [-126.12982731363905, 23.011472701841583],
  [-67.14927405043929, 50.54700547505038]
]  as const;
```
Util then you are probably going to have to use a type assertion:
```ts
const usa_box = [
  [-126.12982731363905, 23.011472701841583] as LngLatLike,
  [-67.14927405043929, 50.54700547505038] as LngLatLike
] as LngLatBoundsLike;
```

Or a builder  function 

```ts


function makeBounds(bounds: LngLatBoundsLike) { return bounds }
const usa_box = makeBounds([
  [-126.12982731363905, 23.011472701841583],
  [-67.14927405043929, 50.54700547505038] 
])
```

**Comment by ekulabuhov** [comment]
Thanks! Using `as const` looks redundant but kinda makes sense. 

I wonder what changed? It didn't complain in v2.9.2 (which i'm upgrading from).

**Comment by j-oliveras** [comment]
@ekulabuhov probably is caused by [Tuples in rest parameters and spread expressions](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#tuples-in-rest-parameters-and-spread-expressions) in version 3.0 or [Mapped types on tuples and arrays](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#mapped-types-on-tuples-and-arrays) in version 3.1.

**Comment by nmain** [comment]
In current versions of typescript, one of the cleaner ways to resolve this is to provide a contextual type:

```ts
const usa_bbox: LngLatBoundsLike = [
  [-126.12982731363905, 23.011472701841583],
  [-67.14927405043929, 50.54700547505038]
];
```
This still does a real typecheck, unlike a type assertion.
